uc-catalog/src/lib.rs

```diff
@@ -0,0 +1,195 @@
+    ) -> Result<Arc<Snapshot>, Box<dyn std::error::Error>> {
+        self.load_snapshot_inner(table_id, table_uri, None, engine)
+            .await
+    }
+
+    /// Load a snapshot of a Delta Table identified by `table_id` and `table_uri` for a specific
+    /// version. Generally, a separate `get_table` call can be used to resolve the table id/uri from
+    /// the table name.
+    pub async fn load_snapshot_at(
+        &self,
+        table_id: &str,
+        table_uri: &str,
+        version: Version,
+        engine: &dyn Engine,
+    ) -> Result<Arc<Snapshot>, Box<dyn std::error::Error>> {
+        self.load_snapshot_inner(table_id, table_uri, Some(version), engine)
+            .await
+    }
+
+    pub(crate) async fn load_snapshot_inner(
+        &self,
+        table_id: &str,
+        table_uri: &str,
+        version: Option<Version>,
+        engine: &dyn Engine,
+    ) -> Result<Arc<Snapshot>, Box<dyn std::error::Error>> {
+        let table_uri = table_uri.to_string();
+        let req = CommitsRequest {
+            table_id: table_id.to_string(),
+            table_uri: table_uri.clone(),
+            start_version: Some(0),
+            end_version: version.and_then(|v| v.try_into().ok()),
+        };
+        // TODO: does it paginate?
+        let commits = self.client.get_commits(req).await?;
+
+        // if commits are present, we ensure they are sorted+contiguous
+        if let Some(commits) = &commits.commits {
+            if !commits.windows(2).all(|w| w[1].version == w[0].version + 1) {
+                return Err("Received non-contiguous commit versions".into());
+            }
+        }
+
+        // we always get back the latest version from commits response, and pass that in to
+        // kernel's Snapshot builder. basically, load_table for the latest version always looks
+        // like a time travel query since we know the latest version ahead of time.
+        //
+        // note there is a weird edge case: if the table was just created it will return
+        // latest_table_version = -1, but the 0.json will exist in the _delta_log.
+        let version: Version = match version {
+            Some(v) => v,
+            None => match commits.latest_table_version {
+                -1 => 0,
+                i => i.try_into()?,
+            },
+        };
+
+        // consume uc-client's Commit and hand back a delta_kernel LogPath
+        let table_url = Url::parse(&table_uri)?;
+        let commits: Vec<_> = commits
+            .commits
+            .unwrap_or_default()
+            .into_iter()
+            .map(|c| -> Result<LogPath, Box<dyn std::error::Error>> {
+                LogPath::staged_commit(
+                    table_url.clone(),
+                    &c.file_name,
+                    c.file_modification_timestamp,
+                    c.file_size.try_into()?,
+                )
+                .map_err(|e| e.into())
+            })
+            .try_collect()?;
```

Comment:
I would like for us to wrap these `?` with an error that explains that we "failed to load snapshot". Otherwise at the top level we just get `Failed to parse staged commit`.

Reply:
If not wrapping, then at least an `error!("Failed while loading snapshot: {e}")`